### ISSUE TITLE
Accessibility Issues for AG Grid

### ISSUE THREAD

**Issue by BhagyashriDhanak-web** [ag-grid/ag-grid#6218]
Hi Team,

When I was doing accessibility testing through axe DevTools on my local system, then I found two critical issues there.

Issues:
1.Certain ARIA roles must contain particular children.
2.ARIA attribute must conform to valid name.

**Repro Steps**
1.Open this [URL](https://plnkr.co/edit/cAwskRX7geMbabi1?open=main.js&preview)
2.Right click and click on Inspect option.
3.Test the page through axe DevTools. you will see the critical issues.

Please resolve this issue as soon as possible.

**Comment by kiril-matev** [comment]
Hi @BhagyashriDhanak-web,

Thanks for submitting this well-formed bug report.

Let me respond to the 3 issues that axe DevTools flags here:
**1/ Certain ARIA roles must contain particular children**

AG Grid uses separate containers to display pinned columns on the left and right for the following visual elements:
* column headers
* floating filter row
* grid rows
* top pinned rows
* bottom pinned rows
* sticky rows

These containers are flagged as unexpected by axe DevTools but this is only based on static DOM analysis which assumes the grid isn't interactive. The user can in fact pin columns on left and right and these containers are used to display the cells in the pinned columns, so the presence of these elements isn't incorrect, they're not unnecessary as the test concludes (based on static DOM analysis in the state of the grid when there's no pinned columns). These containers are displayed (no longer hidden as in your example) only when the user uses column pinning on the left or right. That's when these containers are no longer hidden and display normally.

There's no impact for the accessibility of any AG Grid element because of this as AG Grid updates the aria attributes via internal logic to show/hide these pinned column containers when the user pins/unpins a column.

**2/ ARIA attributes must confirm to valid names**

The attribute that gets flagged is aria-description.

aria-description is an aria attribute that's not yet part of the official published version of WCAG but is instead only part of the draft next version of WCAG. However, the aria-description attribute is quite valuable in adding accessibility information and is supported by all modern browsers, so AG Grid has added support for it too. 
 
Please see the page below for this aria attribute and note the infobox on top:
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-description
 
Note: aria-description is still in W3C Editor's Draft for ARIA 1.3. For the time being, continue to use [aria-describedby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby), which has been supported since ARIA 1.1.

This is why the aria-description attribute is correct and is used by screen readers such as NVDA but is still marked as invalid by automated WCAG audit tools which base their analysis on the official version of WCAG.
 
When the currently draft version of WCAG is published, this will no longer be reported as an issue by automated WCAG audit tools.

**3/ Scrollable region must have keyboard access**

This is an expected failure but without any actual impact. Let me explain. AG Grid uses row virtualization to allow working with large sets of data without creating separate DOM elements to represent each record in the data. For this purpose, AG Grid uses a virtualized container (div element with class=ag-body-viewport) to host the row elements representing rows.  
 
However, this requires the nesting of 2 row containers to implement this virualization in practice. This is why the div with class=ag-body-viewport hosts another container, which actually hosts the row elements, which is actually keyboard navigable.
 
So basically you have the following DOM element structure inside the grid:
<ag-grid>
....
<top-level-viewport> (not keyboard navigable)
    <virtualized rows viewport> (keyboard navigable)
           <row 1>
           ....
           <row N>
 
According to the WCAG standard the top-level viewport contents should be accessible by keyboard. However, the top-level viewport contains another container for the rows. This is why you're getting an accessibility failure in the top-level viewport. This is because the WCAG accessibility standard wasn't built with virtual containers in mind. 
 
However, this accessibility failure is only based on static DOM analysis and has no practical implications for the accessibility of the grid as AG Grid implements its own custom keyboard navigation which ensures all its row elements are keyboard navigable.

As you can see from my responses above, these issues do not actually cause any gaps in accessibility support in AG Grid.

Thank you for your time and for using AG Grid!

**Comment by xianshenglu** [comment]
In your [demo](https://www.ag-grid.com/examples/column-definitions/column-fields/packages/angular/index.html) page, if testing with axeTools there's a new issue:

![image](https://github.com/ag-grid/ag-grid/assets/23273077/c1bee5b9-6f70-4a2a-8792-49189376c8f5)

**Comment by sarea** [comment]
For the `Scrollable region must have keyboard access`
I did this:
```
  onFirstDataRendered() {
    document.querySelectorAll('.ag-center-cols-container').forEach(element => {
      element.setAttribute('tabindex', '0');
    });
  }
```

**Comment by naveen700** [comment]
**ERROR**: Element has children which are not allowed: [role=presentation]

I am facing similar issue and as per suggestion here (https://github.com/dequelabs/axe-core/issues/4419) I removed the tabindex as -1 from the child component within browser's dev tool itself, the issue does not come up in accessibility tool again. My problem is the table has been created with AG-GRID and it creates the whole structure on its own, the internal elements too. I am not able to change the role of internal elements or remove the tabIndex. 

Did you guys have any solution for this or any suggestion ? Any help would be great! Thanks.

![Image](https://github.com/user-attachments/assets/b61886b5-3609-4f5d-97e6-572f7cd5a218)

Removing tabIndex as per this suggestion (If you have any other resolution for this, that would be helpful).

![Image](https://github.com/user-attachments/assets/112f5ede-57b9-4d7b-9c99-3eeceff4358a)

I am using versions below
    "ag-grid-community": "^22.1.1",
    "ag-grid-react": "^22.1.1",
   "node-version" : 16

**Comment by AG-Zoheil** [comment]
Hi 

Thank you for your comment.

I can see you have mentioned you're using AG-Grid version 22, which is quite old. Our current latest version is 33.

Version 22 is before we made significant improvements in accessibility to resolve issues like this.

You can go through our documentation archive to find out which version exactly fixed this, however I would highly recommend upgrading to the latest version of AG-Grid in your application so that you can benefit from all the improvements in functionality, stability, theming and performance we've made since the version you're using.

Kind regards,
Zoheil

**Comment by naveen700** [comment]
Hi @AG-Zoheil ,

Thanks for your quick response. Since this was a legacy project I could not update it to version 33 as it stopped working and giving errors, but I slowly turned down version from 33 to lesser versions and the version which is working fine for me is 29 because in 30 there was change in the way it worked. So 29 is the version which resolved the accessibility issue.

Thank you for your help! Greatly appreciated.
